### PR TITLE
Array fields support

### LINKED ISSUE #77: Support array fields
Current client API does not allow you to specify field schema for `Array` types.

I propose extending the C# API to include support for defining array field schemas, which would allow developers to specify the type of elements in the array along with constraints such as `capacity` and `maxLength`. Below are proposed methods to add to the API for defining array fields with both varchar and scalar types:

```
FieldSchema.CreateVarcharArray("field_id", capacity: 200, maxLength: 255, description: "Array of varchars"),
FieldSchema.CreateArray<float>("field_id", capacity: 200, description: "Array of scalar types")
```

### PR DESCRIPTION
**Pull request by BlackGad**
This PR introduces support for [Array](https://milvus.io/docs/array_data_type.md#Array) fields, encompassing schema definitions as well as data requests, inserts, and upserts.

I have authored a series of unit tests to cover various array types.

Additionally, the Proto submodule has been upgraded to [v2.4](https://github.com/milvus-io/milvus-proto/tree/2.4).

Fixes #77

### CODE REVIEW

File: Milvus.Client/FieldSchema.cs [hunk 1]
```diff
@@ -95,6 +95,42 @@ public static FieldSchema CreateBinaryVector(string name, int dimension, string
...
+    /// <summary>
+    /// Create a field schema for a <c>array</c> of <c>TData</c> field.
+    /// </summary>
+    /// <typeparam name="TData">Determines the element data type of array stored in the field based.</typeparam>
+    /// <param name="name">The field name.</param>
+    /// <param name="capacity">Maximum number of elements that an array field can contain.</param>
+    /// <param name="description">An optional description for the field.</param>
+    public static FieldSchema CreateArray<TData>(
+        string name,
+        int capacity,
```
Comment (roji) [inline Milvus.Client/FieldSchema.cs:107]: Maybe `maxCapacity` as the parameter is actually called in Milvus?

File: Milvus.Client/ArrayFieldData.cs [hunk 2]
```diff
@@ -0,0 +1,187 @@
...
+        };
+
+        foreach (var array in Data)
+        {
+            switch (ElementType)
+            {
+                case MilvusDataType.Bool:
+                    Grpc.BoolArray boolData = new();
+                    boolData.Data.AddRange(array as IEnumerable<bool>);
+                    arrayArray.Data.Add(new ScalarField {BoolData = boolData});
```
Comment (roji) [inline Milvus.Client/ArrayFieldData.cs:58]: nit (below as well):

```suggestion
                    arrayArray.Data.Add(new ScalarField { BoolData = boolData });
```

File: Milvus.Client/FieldData.cs [hunk 3]
```diff
@@ -120,22 +120,40 @@ internal static FieldData FromGrpcFieldData(Grpc.FieldData fieldData)
                 }
 
             case Grpc.FieldData.FieldOneofCase.Scalars:
-                return fieldData.Scalars.DataCase switch
+                return fieldData.Scalars switch
                 {
-                    Grpc.ScalarField.DataOneofCase.BoolData
+                    {DataCase: ScalarField.DataOneofCase.BoolData}
```
Comment (roji) [inline Milvus.Client/FieldData.cs:125]: nit (below as well):

```suggestion
                    { DataCase: ScalarField.DataOneofCase.BoolData }
```

File: Milvus.Client/ArrayFieldData.cs [hunk 4]
```diff
@@ -0,0 +1,187 @@
...
+                case MilvusDataType.None:
+                    throw new MilvusException($"ElementType Error:{DataType}");
+                default:
+                    throw new MilvusException($"ElementType Error:{DataType}, not supported");
+            }
+        }
+
+        return fieldData;
+
+        /*
```
Comment (roji) [inline Milvus.Client/ArrayFieldData.cs:126]: Commented-out code accidentally left here? Should be removed?

File: Milvus.Client/ArrayFieldData.cs [hunk 5]
```diff
@@ -0,0 +1,187 @@
...
+                case MilvusDataType.Json:
+                    Grpc.JSONArray jsonData = new();
+                    var enumerable = array as IEnumerable<string> ?? Enumerable.Empty<string>();
+                    jsonData.Data.AddRange(enumerable.Select(ByteString.CopyFromUtf8));
+                    arrayArray.Data.Add(new ScalarField {JsonData = jsonData});
+                    break;
+                case MilvusDataType.None:
+                    throw new MilvusException($"ElementType Error:{DataType}");
+                default:
+                    throw new MilvusException($"ElementType Error:{DataType}, not supported");
```
Comment (roji) [inline Milvus.Client/ArrayFieldData.cs:120]:
```suggestion

                case MilvusDataType.None:
                    throw new MilvusException($"ElementType Error: {DataType}");

                default:
                    throw new MilvusException($"ElementType Error: {DataType}, not supported");
```

**Review by roji — CHANGES_REQUESTED** [review]
Thanks, that looks pretty good! See review comments mainly for some nits.

BTW can I ask you to submit the 2.4 proto change in a separate PR, as I'm planning on backporting the array support to 2.3 as well?

File: Milvus.Client/ArrayFieldData.cs [hunk 6]
```diff
@@ -0,0 +1,187 @@
...
+public sealed class ArrayFieldData<TElementData> : FieldData<IReadOnlyList<TElementData>>
+{
+    /// <summary>
+    /// Construct an array field
+    /// </summary>
+    /// <param name="fieldName"></param>
+    /// <param name="data"></param>
+    /// <param name="isDynamic"></param>
+    public ArrayFieldData(string fieldName, IEnumerable<IEnumerable<TElementData>> data, bool isDynamic)
+        : base(fieldName, data.Select(x => x.ToArray()).ToArray(), MilvusDataType.Array, isDynamic)
```
[the remark this answers is not in the document]
Comment (BlackGad) [reply Milvus.Client/ArrayFieldData.cs:15]: Must be [supported](https://milvus.io/docs/enable-dynamic-field.md)

> Dynamic schema also provides flexibility in data processing, enabling users to store and retrieve complex data structures in their collections. This includes nested data, **arrays**, and other complex data types.

File: Milvus.Client/ArrayFieldData.cs [hunk 7]
```diff
@@ -0,0 +1,187 @@
...
+/// </summary>
+public sealed class ArrayFieldData<TElementData> : FieldData<IReadOnlyList<TElementData>>
+{
+    /// <summary>
+    /// Construct an array field
+    /// </summary>
+    /// <param name="fieldName"></param>
+    /// <param name="data"></param>
+    /// <param name="isDynamic"></param>
+    public ArrayFieldData(string fieldName, IEnumerable<IEnumerable<TElementData>> data, bool isDynamic)
```
[the remark this answers is not in the document]
Comment (BlackGad) [reply Milvus.Client/ArrayFieldData.cs]: Changed to IReadOnlyList<IReadOnlyList<TElementData>>

File: Milvus.Client/FieldSchema.cs (hunk 1 above)
Comment (BlackGad) [reply Milvus.Client/FieldSchema.cs]: Renamed

File: Milvus.Client/ArrayFieldData.cs (hunk 2 above)
Comment (BlackGad) [reply Milvus.Client/ArrayFieldData.cs]: Changed code style

File: Milvus.Client/ArrayFieldData.cs (hunk 4 above)
Comment (BlackGad) [reply Milvus.Client/ArrayFieldData.cs]: Sorry :)

File: Milvus.Client/ArrayFieldData.cs (hunk 5 above)
Comment (BlackGad) [reply Milvus.Client/ArrayFieldData.cs]: Changed code style

File: Milvus.Client/FieldData.cs (hunk 3 above)
Comment (BlackGad) [reply Milvus.Client/FieldData.cs:125]: Code style changed

**Comment by BlackGad** [comment]
Proto version change reverted. from this PR Added https://github.com/milvus-io/milvus-sdk-csharp/pull/82 for this change

**Comment by BlackGad** [comment]
Did not resolve all the conversation here. Can you take a look for modifications?

File: Milvus.Client/ArrayFieldData.cs (hunk 6 above)
[the remark this answers is not in the document]
Comment (roji) [reply Milvus.Client/ArrayFieldData.cs]: Thanks for confirming,

File: Milvus.Client/ArrayFieldData.cs (hunk 4 above)
Comment (roji) [reply Milvus.Client/ArrayFieldData.cs]: No problem :)

**Review by roji — APPROVED** [review]
Thanks for the quality PR @BlackGad!

**Comment by roji** [comment]
Backported to 2.3.0-preview.2 via 6aea1422e8a31a9fca71143db140b0746fef238d

**Pull request merged**